shinylive/_main.py

```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+
+APP is the path to the primary Shiny app file.
+
+FILES are additional supporting files or directories for the app.
+
+On macOS, you can copy the URL to the clipboard with:
+
+    shinylive url encode app.py | pbcopy
+""",
+)
+@click.option(
+    "-m",
+    "--mode",
+    type=click.Choice(["editor", "app"]),
+    required=True,
+    default="editor",
+    help="The shinylive mode: include the editor or show only the app.",
+)
+@click.option(
+    "-l",
+    "--language",
+    type=click.Choice(["python", "py", "R", "r"]),
+    required=False,
+    default=None,
+    help="The primary language used to run the app, by default inferred from the app file.",
+)
+@click.option(
+    "-v", "--view", is_flag=True, default=False, help="Open the link in a browser."
+)
+@click.option(
+    "--no-header", is_flag=True, default=False, help="Hide the Shinylive header."
+)
+@click.argument("app", type=str, nargs=1, required=True, default="-")
+@click.argument("files", type=str, nargs=-1, required=False)
+def encode(
+    app: str,
+    files: Optional[tuple[str, ...]] = None,
+    mode: str = "editor",
+    language: Optional[str] = None,
+    no_header: bool = False,
+    view: bool = False,
+) -> None:
+    if app == "-":
+        app_in = sys.stdin.read()
+    else:
+        app_in = app
+
+    url = encode_shinylive_url(
+        app=app_in,
+        files=files,
+        mode=mode,
+        language=language,
+        header=not no_header,
+    )
+
+    print(url)
+
+    if view:
+        import webbrowser
+
+        webbrowser.open(url)
+    return
+
+
+@url.command(
+    short_help="Decode a shinylive.io URL.",
+    help="""
+Decode a shinylive.io URL.
+
+URL is the shinylive editor or app URL. If not specified, the URL will be read from
+stdin, allowing you to read the URL from a file or the clipboard.
+
+On macOS, you can read the URL from the clipboard with:
+
+    pbpaste | shinylive url decode
+""",
+)
+@click.option(
+    "--dir",
+    type=str,
+    default=None,
+    help="Output directory into which the app's files will be written. The directory is created if it does not exist. ",
+)
+@click.option(
+    "--json",
+    is_flag=True,
+    default=False,
+    help="Prints the decoded shinylive bundle as JSON to stdout, ignoring --out.",
+)
+@click.argument("url", type=str, nargs=1, default="-")
+def decode(url: str, dir: Optional[str] = None, json: bool = False) -> None:
+    if url == "-":
+        url_in = sys.stdin.read()
+    else:
+        url_in = url
+    bundle = decode_shinylive_url(str(url_in))
+
+    if json:
+        print_as_json(bundle)
+        return
+
+    if dir is not None:
+        import os
+
+        out_dir = Path(dir)
+        os.makedirs(out_dir, exist_ok=True)
+        for file in bundle:
+            with open(out_dir / file["name"], "w") as f_out:
+                f_out.write(
+                    file["content"].encode("utf-8", errors="ignore").decode("utf-8")
+                )
```

Comment:
For binary files, I think this will output base64-encoded text. I believe it needs to be base64-decoded.

(If I'm wrong about it outputting base64-encoded text, then this code would be fine but the corresponding code below, when printing to console, would be wrong -- they can't both be right.)

Reply:
You're totally right, I missed this case. I'll update the command to write the base64-decoded binary content to the output file.

Reply:
2da9c8ab460ae8ef3f24f6e0e0b815e2a74c1ca4